File: sdk/basin-dotnet/src/Basin/Realtime.cs
using System.Net.WebSockets;
using System.Runtime.CompilerServices;
using System.Text;
using System.Text.Json;

namespace Basin;

/// <summary>
/// Options for subscribing to a realtime channel.
/// </summary>
public sealed class SubscribeOptions
{
    /// <summary>
    /// Optional server-side SQL predicate, e.g. "NEW.status = 'paid'".
    /// Invalid predicates arrive as <see cref="RealtimeErrorFrame"/> with code "invalid_filter".
    /// </summary>
    public string? Filter { get; init; }

    /// <summary>
    /// Reconnect cursor — the last seq the client successfully processed.
    /// When given, the server replays missed events from its replay ring,
    /// or emits a <see cref="RealtimeGapFrame"/> if the ring has been evicted.
    /// </summary>
    public long? LastEventId { get; init; }
}

/// <summary>
/// Async WebSocket client for GET /realtime/v1/ws/:project.
///
/// Protocol source: crates/basin-realtime/src/ws.rs (ClientMsg / ServerMsg).
///
/// Auth: Sec-WebSocket-Protocol: basin-v1,&lt;token&gt; subprotocol header.
///
/// Reconnect: exponential backoff (0.5 s → 30 s) with automatic resubscription
/// of all active channels after reconnect.
///
/// Obtain via <see cref="BasinClient.Realtime"/> — do not construct directly.
/// </summary>
public sealed class RealtimeClient : IAsyncDisposable
{
    private readonly string _baseUrl;
    private readonly Func<CancellationToken, Task<string?>> _getToken;
    private readonly Func<string?> _getProjectId;
    // Factory: (url, subprotocols) → already-connected ClientWebSocket.
    // Used in tests to inject a FakeWebSocket without opening a real network connection.
    private readonly Func<string, IEnumerable<string>, ClientWebSocket>? _wsFactory;

    private ClientWebSocket? _ws;
    private CancellationTokenSource? _recvCts;
    private Task? _recvTask;

    private readonly SemaphoreSlim _connectLock = new(1, 1);
    private volatile bool _closed;

    // Active subscriptions: table → options
    private readonly Dictionary<string, SubscribeOptions> _subscriptions = new();

    // Per-table + per-channel queues
    private readonly Dictionary<string, List<System.Threading.Channels.Channel<ServerFrame?>>> _channels = new();

    // Pending subscribe acks: table → TaskCompletionSource
    private readonly Dictionary<string, TaskCompletionSource<bool>> _pendingSubscribed = new();
    private readonly Dictionary<string, TaskCompletionSource<bool>> _pendingUnsubscribed = new();

    // Reconnect backoff params
    private const double BackoffBase = 0.5;
    private const double BackoffMax = 30.0;
    private const double BackoffFactor = 2.0;

    internal RealtimeClient(
        string baseUrl,
        Func<CancellationToken, Task<string?>> getToken,
        Func<string?> getProjectId,
        Func<string, IEnumerable<string>, ClientWebSocket>? wsFactory = null)
    {
        _baseUrl = baseUrl;
        _getToken = getToken;
        _getProjectId = getProjectId;
        _wsFactory = wsFactory;
    }

    /// <summary>True when the WebSocket connection is currently open.</summary>
    public bool IsConnected => _ws?.State == WebSocketState.Open;

    // ------------------------------------------------------------------
    // Connection management
    // ------------------------------------------------------------------

    /// <summary>Open the WebSocket connection. No-op if already connected.</summary>
    public async Task ConnectAsync(CancellationToken ct = default)
    {
        await _connectLock.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            if (_ws?.State == WebSocketState.Open)
                return;

            await DoConnectAsync(ct).ConfigureAwait(false);
        }
        finally
        {
            _connectLock.Release();
        }
    }

    private async Task DoConnectAsync(CancellationToken ct)
    {
        var project = _getProjectId()
            ?? throw new BasinApiException(
                "E_INVALID_REQUEST",
                "realtime requires a project id — pass ProjectId to BasinClient.Builder",
                0);

        var token = await _getToken(ct).ConfigureAwait(false)
            ?? throw new BasinApiException(
                "E_UNAUTHENTICATED",
                "realtime requires a JWT or API key",
                0);

        // Build ws:// URL from http:// base.
        var wsBase = _baseUrl.TrimEnd('/');
        wsBase = wsBase
            .Replace("https://", "wss://", StringComparison.OrdinalIgnoreCase)
            .Replace("http://", "ws://", StringComparison.OrdinalIgnoreCase);

        var url = $"{wsBase}/realtime/v1/ws/{project}";

        // Auth: Sec-WebSocket-Protocol: basin-v1,<token>
        // This is the same subprotocol form used by JS + Python SDKs.
        var subprotocols = new[] { "basin-v1", token };

        ClientWebSocket ws;
        if (_wsFactory is not null)
        {
            // Test injection: factory returns an already-connected socket.
            ws = _wsFactory(url, subprotocols);
        }
        else
        {
            ws = new ClientWebSocket();
            foreach (var proto in subprotocols)
                ws.Options.AddSubProtocol(proto);

            await ws.ConnectAsync(new Uri(url), ct).ConfigureAwait(false);
        }

        _ws = ws;
        _recvCts = new CancellationTokenSource();
        _recvTask = Task.Run(() => RecvLoopAsync(_recvCts.Token), _recvCts.Token);
    }

    /// <summary>Close the WebSocket connection and stop all active listen() generators.</summary>
    public async Task DisconnectAsync(CancellationToken ct = default)
    {
        _closed = true;

        if (_recvCts is not null)
        {
            await _recvCts.CancelAsync().ConfigureAwait(false);
            try { await (_recvTask ?? Task.CompletedTask).ConfigureAwait(false); }
            catch (OperationCanceledException) { }
            _recvCts.Dispose();
            _recvCts = null;
            _recvTask = null;
        }

        await CloseWsAsync().ConfigureAwait(false);

        // Drain all channels with a null sentinel so IAsyncEnumerable consumers exit.
        DrainChannels();
    }

    private async Task CloseWsAsync()
    {
        if (_ws is null)
            return;

        try
        {
            if (_ws.State == WebSocketState.Open)
                await _ws.CloseAsync(WebSocketCloseStatus.NormalClosure, "disconnect", CancellationToken.None)
                          .ConfigureAwait(false);
        }
        catch { /* ignore close errors */ }

        _ws.Dispose();
        _ws = null;
    }

    private void DrainChannels()
    {
        foreach (var chs in _channels.Values)
            foreach (var ch in chs)
                ch.Writer.TryComplete();
    }

    // ------------------------------------------------------------------
    // Send
    // ------------------------------------------------------------------

    private async Task SendAsync(object msg, CancellationToken ct)
    {
        if (_ws is null || _ws.State != WebSocketState.Open)
            throw new BasinApiException("E_UNKNOWN", "realtime socket is not open", 0);

        var bytes = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(msg, BasinJsonOptions.Default));
        // Use the ArraySegment overload so FakeWebSocket can intercept in tests.
        await _ws.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, ct)
                 .ConfigureAwait(false);
    }

    // ------------------------------------------------------------------
    // Receive loop
    // ------------------------------------------------------------------

    private async Task RecvLoopAsync(CancellationToken ct)
    {
        var buffer = new ArraySegment<byte>(new byte[32 * 1024]);
        var sb = new StringBuilder();

        try
        {
            while (!ct.IsCancellationRequested && _ws?.State == WebSocketState.Open)
            {
                sb.Clear();
                WebSocketReceiveResult result;

                do
                {
                    // Use the ArraySegment overload so FakeWebSocket can override it in tests.
                    result = await _ws.ReceiveAsync(buffer, ct).ConfigureAwait(false);

                    if (result.MessageType == WebSocketMessageType.Close)
                        return;

                    if (result.MessageType == WebSocketMessageType.Text)
                        sb.Append(Encoding.UTF8.GetString(buffer.Array!, 0, result.Count));

                } while (!result.EndOfMessage);

                var raw = sb.ToString();
                var frame = ParseFrame(raw);
                if (frame is not null)
                    DispatchFrame(frame);
            }
        }
        catch (OperationCanceledException) { }
        catch (WebSocketException) { }
        catch { /* swallow unexpected recv errors */ }
        finally
        {
            if (_ws is not null)
            {
                _ws.Dispose();
                _ws = null;
            }
        }

        // Unexpected disconnect — reconnect unless closed by caller.
        if (!_closed)
            _ = Task.Run(ReconnectAsync, CancellationToken.None);
    }

    private static ServerFrame? ParseFrame(string raw)
    {
        try
        {
            using var doc = JsonDocument.Parse(raw);
            var root = doc.RootElement;

            if (!root.TryGetProperty("type", out var typeProp) ||
                typeProp.ValueKind != JsonValueKind.String)
                return null;

            var type = typeProp.GetString()!;

            return type switch
            {
                "event"          => Wrap(type, ev: JsonSerializer.Deserialize<RealtimeEvent>(raw, BasinJsonOptions.Default)),
                "error"          => Wrap(type, err: JsonSerializer.Deserialize<RealtimeErrorFrame>(raw, BasinJsonOptions.Default)),
                "gap"            => Wrap(type, gap: JsonSerializer.Deserialize<RealtimeGapFrame>(raw, BasinJsonOptions.Default)),
                "subscribed"     => Wrap(type, sub: JsonSerializer.Deserialize<SubscribedFrame>(raw, BasinJsonOptions.Default)),
                "unsubscribed"   => Wrap(type, unsub: JsonSerializer.Deserialize<UnsubscribedFrame>(raw, BasinJsonOptions.Default)),
                "presence_state" => Wrap(type, ps: JsonSerializer.Deserialize<PresenceStateFrame>(raw, BasinJsonOptions.Default)),
                "presence_diff"  => Wrap(type, pd: JsonSerializer.Deserialize<PresenceDiffFrame>(raw, BasinJsonOptions.Default)),
                // "presenceerror" — no underscore, from Rust rename_all="lowercase" on PresenceError
                "presenceerror"  => Wrap(type, pe: JsonSerializer.Deserialize<PresenceErrorFrame>(raw, BasinJsonOptions.Default)),
                _                => null,
            };
        }
        catch
        {
            return null;
        }
    }

    private static ServerFrame Wrap(string type,
        RealtimeEvent? ev = null,
        RealtimeErrorFrame? err = null,
        RealtimeGapFrame? gap = null,
        SubscribedFrame? sub = null,
        UnsubscribedFrame? unsub = null,
        PresenceStateFrame? ps = null,
        PresenceDiffFrame? pd = null,
        PresenceErrorFrame? pe = null) =>
        new()
        {
            Type = type,
            Event = ev,
            Error = err,
            Gap = gap,
            Subscribed = sub,
            Unsubscribed = unsub,
            PresenceState = ps,
            PresenceDiff = pd,
            PresenceError = pe,
        };

    private void DispatchFrame(ServerFrame frame)
    {
        switch (frame.Type)
        {
            case "event":
            {
                var table = frame.Event!.Table;
                RouteToChannels(table, frame);
                break;
            }
            case "error":
            {
                RouteToChannels(frame.Error!.Table, frame);
                break;
            }
            case "gap":
            {
                RouteToChannels(frame.Gap!.Table, frame);
                break;
            }
            case "subscribed":
            {
                var table = frame.Subscribed!.Table;
                if (_pendingSubscribed.TryGetValue(table, out var tcs))
                {
                    _pendingSubscribed.Remove(table);
                    tcs.TrySetResult(true);
                }
                break;
            }
            case "unsubscribed":
            {
                var table = frame.Unsubscribed!.Table;
                if (_pendingUnsubscribed.TryGetValue(table, out var tcs))
                {
                    _pendingUnsubscribed.Remove(table);
                    tcs.TrySetResult(true);
                }
                break;
            }
            case "presence_state":
                RouteToChannels(frame.PresenceState!.Channel, frame);
                break;
            case "presence_diff":
                RouteToChannels(frame.PresenceDiff!.Channel, frame);
                break;
            case "presenceerror":
                RouteToChannels(frame.PresenceError!.Channel, frame);
                break;
        }
    }

    private void RouteToChannels(string key, ServerFrame frame)
    {
        if (!_channels.TryGetValue(key, out var chs))
            return;

        foreach (var ch in chs)
            ch.Writer.TryWrite(frame);
    }

    // ------------------------------------------------------------------
    // Reconnect
    // ------------------------------------------------------------------

    private async Task ReconnectAsync()
    {
        var attempt = 0;
        while (!_closed)
        {
            var delay = Math.Min(BackoffBase * Math.Pow(BackoffFactor, attempt), BackoffMax);
            await Task.Delay(TimeSpan.FromSeconds(delay)).ConfigureAwait(false);
            attempt++;

            try
            {
                await _connectLock.WaitAsync().ConfigureAwait(false);
                try
                {
                    if (_ws?.State == WebSocketState.Open)
                        return; // another path already reconnected

                    await DoConnectAsync(CancellationToken.None).ConfigureAwait(false);
                }
                finally
                {
                    _connectLock.Release();
                }

                // Re-issue all active subscriptions.
                await ResubscribeAsync().ConfigureAwait(false);
                return;
            }
            catch { /* retry */ }
        }
    }

    private async Task ResubscribeAsync()
    {
        foreach (var (table, opts) in _subscriptions.ToList())
        {
            try
            {
                var tcs = new TaskCompletionSource<bool>();
                _pendingSubscribed[table] = tcs;
                await SendSubscribeAsync(table, opts, CancellationToken.None).ConfigureAwait(false);
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
                await tcs.Task.WaitAsync(cts.Token).ConfigureAwait(false);
            }
            catch { /* log in production */ }
        }
    }

    // ------------------------------------------------------------------
    // Subscribe / unsubscribe
    // ------------------------------------------------------------------

    private Task SendSubscribeAsync(string table, SubscribeOptions opts, CancellationToken ct)
    {
        var msg = new Dictionary<string, object?> { ["type"] = "subscribe", ["table"] = table };
        if (opts.Filter is not null)
            msg["filter"] = opts.Filter;
        if (opts.LastEventId is not null)
            msg["last_event_id"] = opts.LastEventId;

        return SendAsync(msg, ct);
    }

    private async Task SubscribeInternalAsync(string table, SubscribeOptions opts, CancellationToken ct)
    {
        await ConnectAsync(ct).ConfigureAwait(false);
        _subscriptions[table] = opts;

        if (!_channels.ContainsKey(table))
            _channels[table] = new List<System.Threading.Channels.Channel<ServerFrame?>>();

        var tcs = new TaskCompletionSource<bool>();
        _pendingSubscribed[table] = tcs;

        await SendSubscribeAsync(table, opts, ct).ConfigureAwait(false);

        using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        cts.CancelAfter(TimeSpan.FromSeconds(30));
        await tcs.Task.WaitAsync(cts.Token).ConfigureAwait(false);
    }

    private async Task UnsubscribeInternalAsync(string table, CancellationToken ct)
    {
        if (_ws?.State != WebSocketState.Open)
            return;

        var tcs = new TaskCompletionSource<bool>();
        _pendingUnsubscribed[table] = tcs;

        await SendAsync(new { type = "unsubscribe", table }, ct).ConfigureAwait(false);

        try
        {
            using var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
            cts.CancelAfter(TimeSpan.FromSeconds(10));
            await tcs.Task.WaitAsync(cts.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) { /* timeout — continue */ }

        _subscriptions.Remove(table);
    }

    // ------------------------------------------------------------------
    // Public API — IAsyncEnumerable
    // ------------------------------------------------------------------

    /// <summary>
    /// Subscribe to change events for <paramref name="table"/> and yield
    /// <see cref="ServerFrame"/> values via an async stream.
    ///
    /// The subscription is established on entry and torn down when the
    /// caller cancels <paramref name="ct"/> or disposes the enumerator.
    /// Multiple concurrent Listen() calls for the same table share one
    /// subscription.
    ///
    /// Yields event / error / gap / presence_state / presence_diff / presenceerror frames.
    /// </summary>
    public async IAsyncEnumerable<ServerFrame> ListenAsync(
        string table,
        SubscribeOptions? options = null,
        [EnumeratorCancellation] CancellationToken ct = default)
    {
        var opts = options ?? new SubscribeOptions();
        var channel = System.Threading.Channels.Channel.CreateUnbounded<ServerFrame?>(
            new System.Threading.Channels.UnboundedChannelOptions { SingleReader = true });

        if (!_channels.ContainsKey(table))
            _channels[table] = new List<System.Threading.Channels.Channel<ServerFrame?>>();

        _channels[table].Add(channel);

        try
        {
            await SubscribeInternalAsync(table, opts, ct).ConfigureAwait(false);

            await foreach (var frame in channel.Reader.ReadAllAsync(ct))
            {
                if (frame is null)
                    yield break;

                yield return frame;
            }
        }
        finally
        {
            var list = _channels.GetValueOrDefault(table);
            list?.Remove(channel);
            channel.Writer.TryComplete();

            if (list?.Count == 0 && _subscriptions.ContainsKey(table))
                await UnsubscribeInternalAsync(table, CancellationToken.None).ConfigureAwait(false);
        }
    }

    // ------------------------------------------------------------------
    // Presence
    // ------------------------------------------------------------------

    /// <summary>
    /// Send {"type":"presence_track",...} to register in a presence channel.
    /// The server enforces that client_id matches the authenticated JWT identity.
    /// </summary>
    public async Task PresenceTrackAsync(
        string channel,
        string clientId,
        object? metadata = null,
        CancellationToken ct = default)
    {
        await ConnectAsync(ct).ConfigureAwait(false);
        await SendAsync(new
        {
            type = "presence_track",
            channel,
            client_id = clientId,
            metadata = metadata ?? new { },
        }, ct).ConfigureAwait(false);
    }

    /// <summary>Send {"type":"presence_untrack",...}.</summary>
    public async Task PresenceUntrackAsync(
        string channel,
        string clientId,
        CancellationToken ct = default)
    {
        await ConnectAsync(ct).ConfigureAwait(false);
        await SendAsync(new { type = "presence_untrack", channel, client_id = clientId }, ct)
            .ConfigureAwait(false);
    }

    /// <summary>Send {"type":"heartbeat",...} to refresh the presence TTL.</summary>
    public async Task PresenceHeartbeatAsync(
        string channel,
        string clientId,
        CancellationToken ct = default)
    {
        await ConnectAsync(ct).ConfigureAwait(false);
        await SendAsync(new { type = "heartbeat", channel, client_id = clientId }, ct)
            .ConfigureAwait(false);
    }

    // ------------------------------------------------------------------
    // IAsyncDisposable
    // ------------------------------------------------------------------

    public async ValueTask DisposeAsync()
    {
        await DisconnectAsync().ConfigureAwait(false);
        _connectLock.Dispose();
    }
}

File: sdk/basin-swift/Sources/Basin/Realtime.swift
/// Realtime WebSocket client for Basin.
///
/// Connects to GET /realtime/v1/ws/:project and delivers change events via
/// AsyncThrowingStream.
///
/// Protocol source of truth: crates/basin-realtime/src/ws.rs
/// (ClientMsg / ServerMsg + presence.rs serialize_presence_* helpers).
///
/// Auth: Sec-WebSocket-Protocol "basin-v1,<token>" subprotocol header — the
/// server also accepts Authorization: Bearer but browsers and URLSessionWebSocketTask
/// make the subprotocol form simpler for client-originated connections.
///
/// Reconnect: exponential backoff 0.5 s → 1 s → 2 s … capped at 30 s.
/// All active subscriptions are reinstated automatically on reconnect.
/// Pass lastEventID to listen() for server-side replay of missed events.

import Foundation

// MARK: - Frame parsing

private func parseFrame(_ text: String) -> ServerFrame? {
    guard let data = text.data(using: .utf8),
          let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let type = json["type"] as? String else { return nil }

    let frameData = data
    switch type {
    case "event":
        return (try? JSONDecoder().decode(RealtimeEvent.self, from: frameData)).map { .event($0) }
    case "error":
        return (try? JSONDecoder().decode(RealtimeErrorFrame.self, from: frameData)).map { .error($0) }
    case "gap":
        return (try? JSONDecoder().decode(RealtimeGapFrame.self, from: frameData)).map { .gap($0) }
    case "subscribed":
        return (try? JSONDecoder().decode(SubscribedFrame.self, from: frameData)).map { .subscribed($0) }
    case "unsubscribed":
        return (try? JSONDecoder().decode(UnsubscribedFrame.self, from: frameData)).map { .unsubscribed($0) }
    case "presence_state":
        return (try? JSONDecoder().decode(PresenceStateFrame.self, from: frameData)).map { .presenceState($0) }
    case "presence_diff":
        return (try? JSONDecoder().decode(PresenceDiffFrame.self, from: frameData)).map { .presenceDiff($0) }
    case "presenceerror":
        return (try? JSONDecoder().decode(PresenceErrorFrame.self, from: frameData)).map { .presenceError($0) }
    default:
        return nil
    }
}

// MARK: - Backoff

private func backoffDelay(attempt: Int) -> TimeInterval {
    min(0.5 * pow(2.0, Double(attempt)), 30.0)
}

// MARK: - RealtimeClient

/// Multiplexed WebSocket client for GET /realtime/v1/ws/:project.
///
/// Obtain via `BasinClient.realtime` — do not construct directly.
///
/// The primary API is `listen(table:)` which returns an `AsyncThrowingStream`
/// of `ServerFrame` values. For table-change events only, filter with
/// `if case .event(let e) = frame`.
public actor RealtimeClient {
    private let baseURL: String
    private let getToken: @Sendable () async -> String?
    private let getProjectID: @Sendable () async -> String?

    // URLSessionWebSocketTask is not an actor, keep inside the actor.
    private var task: URLSessionWebSocketTask?
    private var closed = false

    // Table → set of continuations that want frames for that table.
    private var tableListeners: [String: [UUID: AsyncThrowingStream<ServerFrame, Error>.Continuation]] = [:]
    // Channel → continuations for presence frames.
    private var channelListeners: [String: [UUID: AsyncThrowingStream<ServerFrame, Error>.Continuation]] = [:]
    // Pending "subscribed" acks: table → continuation to resume.
    private var pendingSubscribed: [String: CheckedContinuation<Void, Error>] = [:]
    // Active subscriptions (for reconnect).
    private var subscriptions: [String: SubscribeOpts] = [:]

    private struct SubscribeOpts {
        let filter: String?
        let lastEventID: Int?
    }

    public init(
        baseURL: String,
        getToken: @escaping @Sendable () async -> String?,
        getProjectID: @escaping @Sendable () async -> String?
    ) {
        self.baseURL = baseURL
        self.getToken = getToken
        self.getProjectID = getProjectID
    }

    // MARK: - Connection

    /// Open the WebSocket connection. No-op if already connected.
    public func connect() async throws {
        guard task == nil else { return }
        try await _doConnect()
    }

    private func _doConnect() async throws {
        guard let projectID = await getProjectID() else {
            throw BasinApiError(rawCode: "E_INVALID_REQUEST",
                                message: "realtime requires a project id — pass projectID to BasinClient",
                                status: 0)
        }
        guard let token = await getToken() else {
            throw BasinApiError(rawCode: "E_UNAUTHENTICATED",
                                message: "realtime requires a JWT or API key",
                                status: 0)
        }

        var wsBase = baseURL
            .replacingOccurrences(of: "https://", with: "wss://")
            .replacingOccurrences(of: "http://", with: "ws://")
        wsBase = wsBase.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let urlStr = "\(wsBase)/realtime/v1/ws/\(projectID)"
        guard let url = URL(string: urlStr) else {
            throw BasinApiError(rawCode: "E_INVALID_REQUEST", message: "invalid realtime URL", status: 0)
        }

        var request = URLRequest(url: url)
        // Auth via subprotocol header (browser-friendly, mirrors Python/JS SDK).
        request.setValue("basin-v1, \(token)", forHTTPHeaderField: "Sec-WebSocket-Protocol")

        let t = URLSession.shared.webSocketTask(with: request)
        self.task = t
        t.resume()

        // Start receive loop.
        Task { await self._receiveLoop() }
    }

    /// Disconnect and close all streams.
    public func disconnect() async {
        closed = true
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
        _drainAll(throwing: CancellationError())
    }

    public var isConnected: Bool { task != nil }

    // MARK: - Receive loop

    private func _receiveLoop() async {
        guard let t = task else { return }
        do {
            while true {
                let message = try await t.receive()
                switch message {
                case .string(let text):
                    if let frame = parseFrame(text) { _dispatch(frame) }
                case .data(let data):
                    if let text = String(data: data, encoding: .utf8),
                       let frame = parseFrame(text) { _dispatch(frame) }
                @unknown default:
                    break
                }
            }
        } catch {
            self.task = nil
            if !closed {
                Task { await self._reconnect() }
            } else {
                _drainAll(throwing: error)
            }
        }
    }

    private func _dispatch(_ frame: ServerFrame) {
        switch frame {
        case .event(let e):
            for cont in tableListeners[e.table]?.values ?? [:].values {
                cont.yield(.event(e))
            }
        case .error(let e):
            for cont in tableListeners[e.table]?.values ?? [:].values {
                cont.yield(.error(e))
            }
        case .gap(let g):
            for cont in tableListeners[g.table]?.values ?? [:].values {
                cont.yield(.gap(g))
            }
        case .subscribed(let s):
            pendingSubscribed[s.table]?.resume()
            pendingSubscribed[s.table] = nil
        case .unsubscribed:
            break
        case .presenceState(let p):
            for cont in channelListeners[p.channel]?.values ?? [:].values {
                cont.yield(.presenceState(p))
            }
        case .presenceDiff(let p):
            for cont in channelListeners[p.channel]?.values ?? [:].values {
                cont.yield(.presenceDiff(p))
            }
        case .presenceError(let p):
            for cont in channelListeners[p.channel]?.values ?? [:].values {
                cont.yield(.presenceError(p))
            }
        }
    }

    private func _drainAll(throwing error: Error) {
        for (_, subs) in tableListeners { for (_, c) in subs { c.finish(throwing: error) } }
        for (_, subs) in channelListeners { for (_, c) in subs { c.finish(throwing: error) } }
        tableListeners = [:]
        channelListeners = [:]
        for (_, cc) in pendingSubscribed { cc.resume(throwing: error) }
        pendingSubscribed = [:]
    }

    // MARK: - Reconnect

    private func _reconnect() async {
        var attempt = 0
        while !closed {
            let delay = backoffDelay(attempt: attempt)
            try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            attempt += 1
            do {
                try await _doConnect()
                try await _resubscribeAll()
                return
            } catch {
                // Continue retry loop.
            }
        }
    }

    private func _resubscribeAll() async throws {
        for (table, opts) in subscriptions {
            try await _sendSubscribe(table: table, opts: opts)
        }
    }

    // MARK: - Subscribe / send

    private func _sendMessage(_ dict: [String: Any]) async throws {
        guard let t = task else {
            throw BasinApiError(rawCode: "E_UNKNOWN", message: "realtime socket is not open", status: 0)
        }
        let data = try JSONSerialization.data(withJSONObject: dict)
        let text = String(data: data, encoding: .utf8) ?? ""
        try await t.send(.string(text))
    }

    private func _sendSubscribe(table: String, opts: SubscribeOpts) async throws {
        var msg: [String: Any] = ["type": "subscribe", "table": table]
        if let f = opts.filter { msg["filter"] = f }
        if let id = opts.lastEventID { msg["last_event_id"] = id }
        try await _sendMessage(msg)
    }

    private func _subscribe(table: String, opts: SubscribeOpts) async throws {
        try await connect()
        subscriptions[table] = opts
        if tableListeners[table] == nil { tableListeners[table] = [:] }
        try await withCheckedThrowingContinuation { (cc: CheckedContinuation<Void, Error>) in
            pendingSubscribed[table] = cc
            Task {
                do {
                    try await _sendSubscribe(table: table, opts: opts)
                } catch {
                    pendingSubscribed[table]?.resume(throwing: error)
                    pendingSubscribed[table] = nil
                }
            }
        }
    }

    private func _unsubscribe(table: String) async throws {
        guard task != nil else { return }
        try await _sendMessage(["type": "unsubscribe", "table": table])
        subscriptions[table] = nil
    }

    // MARK: - Public API

    /// Listen for change events on `table`.
    ///
    /// Returns an `AsyncThrowingStream<ServerFrame, Error>`. The stream yields
    /// `ServerFrame.event`, `.error`, and `.gap` frames for the subscribed table.
    /// The subscription is maintained across reconnects; pass `lastEventID` to
    /// request server-side replay of missed events.
    ///
    /// - Parameter table: Table name to subscribe to.
    /// - Parameter filter: Optional server-side SQL predicate, e.g. `"NEW.status = 'paid'"`.
    /// - Parameter lastEventID: Last seq successfully processed; triggers server-side replay.
    public func listen(
        table: String,
        filter: String? = nil,
        lastEventID: Int? = nil
    ) -> AsyncThrowingStream<ServerFrame, Error> {
        let id = UUID()
        let opts = SubscribeOpts(filter: filter, lastEventID: lastEventID)
        return AsyncThrowingStream { cont in
            Task {
                do {
                    try await _subscribe(table: table, opts: opts)
                    if self.tableListeners[table] == nil { self.tableListeners[table] = [:] }
                    self.tableListeners[table]![id] = cont
                    cont.onTermination = { [weak self] _ in
                        Task {
                            await self?._removeTableListener(table: table, id: id)
                        }
                    }
                } catch {
                    cont.finish(throwing: error)
                }
            }
        }
    }

    private func _removeTableListener(table: String, id: UUID) async {
        tableListeners[table]?[id] = nil
        if tableListeners[table]?.isEmpty == true {
            tableListeners[table] = nil
            try? await _unsubscribe(table: table)
        }
    }

    // MARK: - Presence

    /// Send `{"type":"presence_track",...}` to register in a presence channel.
    public func presenceTrack(channel: String, clientID: String, metadata: [String: Any] = [:]) async throws {
        try await connect()
        try await _sendMessage([
            "type": "presence_track",
            "channel": channel,
            "client_id": clientID,
            "metadata": metadata
        ])
    }

    /// Send `{"type":"presence_untrack",...}`.
    public func presenceUntrack(channel: String, clientID: String) async throws {
        try await connect()
        try await _sendMessage(["type": "presence_untrack", "channel": channel, "client_id": clientID])
    }

    /// Send `{"type":"heartbeat",...}` to refresh the presence TTL.
    public func presenceHeartbeat(channel: String, clientID: String) async throws {
        try await connect()
        try await _sendMessage(["type": "heartbeat", "channel": channel, "client_id": clientID])
    }

    /// Listen for presence frames on `channel`.
    ///
    /// Yields `PresenceStateFrame` (snapshot on join) and `PresenceDiffFrame` (deltas).
    public func listenPresence(channel: String) -> AsyncThrowingStream<ServerFrame, Error> {
        let id = UUID()
        return AsyncThrowingStream { cont in
            Task {
                try await self.connect()
                if self.channelListeners[channel] == nil { self.channelListeners[channel] = [:] }
                self.channelListeners[channel]![id] = cont
                cont.onTermination = { [weak self] _ in
                    Task { await self?._removeChannelListener(channel: channel, id: id) }
                }
            }
        }
    }

    private func _removeChannelListener(channel: String, id: UUID) async {
        channelListeners[channel]?[id] = nil
        if channelListeners[channel]?.isEmpty == true {
            channelListeners[channel] = nil
        }
    }
}
